addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/ModernAppDetectionScanner.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2019 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.pscanrulesAlpha;
+
+import java.util.List;
+import net.htmlparser.jericho.Element;
+import net.htmlparser.jericho.HTMLElementName;
+import net.htmlparser.jericho.Source;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.pscan.PassiveScanThread;
+import org.zaproxy.zap.extension.pscan.PluginPassiveScanner;
+
+/** A class to passively scan responses for indications that this is a modern web application. */
+public class ModernAppDetectionScanner extends PluginPassiveScanner {
+
+    private PassiveScanThread parent = null;
+
+    /** Prefix for internationalized messages used by this rule */
+    private static final String MESSAGE_PREFIX = "pscanalpha.modernapp.";
+
+    /**
+     * gets the name of the scanner
+     *
+     * @return
+     */
+    @Override
+    public String getName() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "name");
+    }
+
+    /**
+     * scans the HTTP request sent (in fact, does nothing)
+     *
+     * @param msg
+     * @param id
+     */
+    @Override
+    public void scanHttpRequestSend(HttpMessage msg, int id) {
+        // do nothing
+    }
+
+    /**
+     * scans the HTTP response for signatures that might indicate this is a modern web application
+     *
+     * @param msg
+     * @param id
+     * @param source unused
+     */
+    @Override
+    public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
+        if (!msg.getResponseHeader().isHtml()) {
+            // Only check HTML responses
+        }
+        String evidence = null;
+        String otherInfo = null;
+
+        List<Element> links = source.getAllElements(HTMLElementName.A);
+        if (links.size() == 0) {
```

Comment:
Could this be null? (I don’t recall the behaviour of getAllElements (Same for the one below)

Reply:
No it cant - the tests confirm this :D